plugins/debuginfo-install.py

```diff
@@ -79,54 +83,95 @@ def run(self):
                 errors_spec.append(pkgspec)
-            for pkg in package_query:
-                self._di_install(pkg)
 
-        if not self.dbgdone:
-            logger.info(_("No debuginfo packages available to install"))
+            # attempt to install debuginfo and debugsource for the highest
+            # listed version of the package (in case the package is installed,
+            # only the installed version is listed)
+            for pkg in packages.values():
+                package = pkg[0]
+
+                if not self._install(package, package.debug_name, solution["nevra"]):
+                    self.debuginfo_missing.add(str(package))
+
+                if not self._install(package, package.debugsource_name, solution["nevra"]):
+                    self.debugsource_missing.add(str(package))
+
+        if self.debuginfo_missing:
+            logger.info(_("Could not find debuginfo packages for the following packages: %s") %
+                        ", ".join(sorted(self.debuginfo_missing)))
+
+        if self.debugsource_missing:
+            logger.info(_("Could not find debugsource packages for the following packages: %s") %
+                        ", ".join(sorted(self.debugsource_missing)))
 
         if errors_spec and self.base.conf.strict:
             raise dnf.exceptions.PackagesNotAvailableError(_("Unable to find a match"),
                                                            pkg_spec=' '.join(errors_spec))
 
-    def _dbg_available(self, dbgname, package, match_evra):
-        if match_evra:
-            return self.packages_available.filter(
-                name=dbgname,
-                epoch=int(package.epoch),
-                version=str(package.version),
-                release=str(package.release),
-                arch=str(package.arch))
-        else:
-            return self.packages_available.filter(
-                name=dbgname,
-                arch=str(package.arch))
-
-    def _di_install(self, package):
-        for dbgname in [package.debug_name, package.source_debug_name]:
-            if dbgname in self.dbgdone:
-                break
-            if self._dbg_available(dbgname, package, True):
-                di = "{0}-{1}:{2}-{3}.{4}".format(
-                    dbgname,
-                    package.epoch,
-                    package.version,
-                    package.release,
-                    package.arch)
-                self.base.install(di)
-            elif self._dbg_available(dbgname, package, False):
-                di = "{0}.{1}".format(dbgname, package.arch)
-                self.base.install(di)
-            else:
-                continue
-            self.dbgdone.append(dbgname)
-            break
-        else:
-            logger.info(
-                _("Could not find debuginfo for package: %s") % package)
+    def _install(self, package, debug_name, base_nevra):
+        if debug_name in self.packages_installed:
```

Comment:
Is this condition really necessary? When I use argument foo-1 and foo-2 and both version will exist, the code will install packages for the first argument. The second argument will be silently skipped. By removal of that condition it will fail with depsolver error.

Reply:
This is one of the few remnants of the original code :slightly_smiling_face: It's clearly supposed to prevent reinstalling the same packages many times over. However, the `self.base.goal` already does that (and better), so I'll drop this.